src/battle/battle_system.c

```diff
@@ -1303,3 +1331,634 @@ u32 CalcMoneyLoss(Party *party, PlayerProfile *profile) {
+            }
+            FillWindowPixelBuffer(window, 0xFF);
+            String_Copy(bsys->msgBuffer, msg);
+            index = AddTextPrinterParameterized(window, 1, bsys->msgBuffer, 0, 0, delay, ov12_0223CF14);
+            String_Delete(msg);
+        } else {
+            MsgData *data;
+            String *msg;
+            int stringId;
+            u32 msgId;
+            int i;
+    
+            if (a2 == 0x64) {
+                stringId = trainerId * 3 + 1;
+            } else {
+                stringId = trainerId * 3 + 2;
+            }
+            
+            for (i = 0; i < 4; i++) {
+                if (PlayerProfile_GetVersion(bsys->playerProfile[i]) == 0) {
+                    break;
+                }
+            }
+    
+            if (i == 4) {
+                msgId = msg_0197_00724;
+            } else {
+                msgId = msg_0197_00723;
+            }
+    
+            data = NewMsgDataFromNarc(MSGDATA_LOAD_DIRECT, NARC_msgdata_msg, msgId, HEAP_ID_BATTLE);
+            msg = NewString_ReadMsgData(data, stringId);
+            FillWindowPixelBuffer(window, 0xFF);
+            String_Copy(bsys->msgBuffer, msg);
+            index = AddTextPrinterParameterized(window, 1, bsys->msgBuffer, 0, 0, delay, ov12_0223CF14);
+            String_Delete(msg);
+            DestroyMsgData(data);
+        }
+    } else {
+        GetTrainerMessageByIdPair(trainerId, a2, bsys->msgBuffer, HEAP_ID_BATTLE);
+        FillWindowPixelBuffer(window, 0xFF);
+        index = AddTextPrinterParameterized(window, 1, bsys->msgBuffer, 0, 0, delay, ov12_0223CF14);
+    }
+    return index;
+}
+
+u32 BattleSystem_PrintBattleMessage(BattleSystem *bsys, MsgData *data, BattleMessage *msg, u8 delay) {
+    Window *window = BattleSystem_GetWindow(bsys, 0);
+    BattleSystem_AdjustMessageForSide(bsys, msg);
+    BattleSystem_BufferMessage(bsys, msg);
+    BattleMessage_ExpandPlaceholders(bsys, data, msg);
+    FillWindowPixelBuffer(window, 0xFF);
+    return AddTextPrinterParameterized(window, 1, bsys->msgBuffer, 0, 0, delay, ov12_0223CF14);
+}
+
+u32 ov12_0223C4E8(BattleSystem *bsys, Window *window, MsgData *data, BattleMessage *msg, int x, int y, int flag, int width, int delay) {
+    int dx;
+    
+    BattleSystem_AdjustMessageForSide(bsys, msg);
+    BattleSystem_BufferMessage(bsys, msg);
+    BattleMessage_ExpandPlaceholders(bsys, data, msg);
+    
+    if (flag & 1) {
+        FillWindowPixelBuffer(window, 0xFF);
+    }
+    
+    if (flag & 2) {
+        dx = width - FontID_String_GetWidth(0, bsys->msgBuffer, 0);  
+    } else {
+        dx = 0;
+    }
+    
+    return AddTextPrinterParameterized(window, 0, bsys->msgBuffer, x + dx, y, delay, ov12_0223CF14);
+}
+
+static void BattleSystem_AdjustMessageForSide(BattleSystem *bsys, BattleMessage *msg) {
+    u32 battleType = BattleSystem_GetBattleType(bsys);
+    
+    if (msg->tag & 0x80) {
```

Comment:
msg tag constants?

Reply:
The last few bits of the message tag are some sorta flag, but I'm not actually sure what those flags are supposed to be atm unless Lhea can enlighten us